distributed/shuffle/tests/test_shuffle.py

```diff
@@ -2309,7 +2337,55 @@ def make_partition(i):
 
 
 @gen_cluster(client=True)
-async def test_set_index_p2p(c, s, *workers):
+async def test_handle_categorical_data(c, s, a, b):
+    """Regression test for https://github.com/dask/distributed/issues/8186"""
+    df = dd.from_dict(
+        {
+            "a": [1, 2, 3, 4, 5],
+            "b": [
+                "x",
+                "y",
+                "x",
+                "y",
+                "z",
+            ],
+        },
+        npartitions=2,
+    )
+    df.b = df.b.astype("category")
+    shuffled = df.shuffle("a", shuffle="p2p")
+    result, expected = await c.compute([shuffled, df], sync=True)
+    dd.assert_eq(result, expected, check_categorical=False)
```

Comment:
There's one caveat here: The expected `dtype` of column `b` is `string`, but after the shuffle, we end up with `object`. There's not much we can do about this at the moment, `meta` also has `object`.

Reply:
Related: dask/dask#6242